gnd/src/main/java/com/google/android/gnd/repository/ProjectRepository.java

```diff
@@ -0,0 +1,159 @@
+import java.util.concurrent.TimeUnit;
+import java8.util.Optional;
+import javax.inject.Inject;
+import javax.inject.Singleton;
+
+/**
+ * Coordinates persistence and retrieval of {@link Project} instances from remote, local, and in
+ * memory data stores. For more details on this pattern and overall architecture, see
+ * https://developer.android.com/jetpack/docs/guide.
+ */
+@Singleton
+public class ProjectRepository {
+  private static final String TAG = ProjectRepository.class.getSimpleName();
+  private static final long LOAD_REMOTE_PROJECT_TIMEOUT_SECS = 5;
+  private static final long LOAD_REMOTE_PROJECT_SUMMARIES_TIMEOUT_SECS = 10;
+
+  private final InMemoryCache cache;
+  private final LocalDataStore localDataStore;
+  private final RemoteDataStore remoteDataStore;
+  private final Flowable<Loadable<Project>> activeProjectStream;
+  private final FlowableProcessor<Optional<String>> activateProjectRequests;
+  private final LocalValueStore localValueStore;
+
+  @Inject
+  public ProjectRepository(
+      LocalDataStore localDataStore,
+      RemoteDataStore remoteDataStore,
+      InMemoryCache cache,
+      LocalValueStore localValueStore) {
+    this.localDataStore = localDataStore;
+    this.remoteDataStore = remoteDataStore;
+    this.cache = cache;
+    this.localValueStore = localValueStore;
+
+    // BehaviorProcessor re-emits last requested project id to late subscribers.
+    this.activateProjectRequests = BehaviorProcessor.create();
+
+    // Load project when requested id changes, caching the last loaded project.
+    this.activeProjectStream =
+        activateProjectRequests
+            .distinctUntilChanged()
+            .doOnNext(id -> Log.v(TAG, "Requested project id changed: " + id))
+            .switchMap(this::onActivateProjectRequest)
+            .replay(1)
+            .refCount();
+  }
+
+  private Flowable<Loadable<Project>> onActivateProjectRequest(Optional<String> projectId) {
+    // Empty id indicates intent to deactivate the current project. Used on sign out.
+    if (projectId.isEmpty()) {
+      return Flowable.just(Loadable.notLoaded());
+    }
+    String id = projectId.get();
+
+    return syncProjectWithRemote(id)
+        .onErrorResumeNext(error -> getProjectFromLocal(id, error))
+        .doOnSuccess(__ -> localValueStore.setLastActiveProjectId(id))
+        .toFlowable()
+        .compose(Loadable::loadingOnceAndWrap);
+  }
+
+  private Single<Project> syncProjectWithRemote(String id) {
+    return remoteDataStore
+        .loadProject(id)
+        .timeout(LOAD_REMOTE_PROJECT_TIMEOUT_SECS, TimeUnit.SECONDS)
+        .doOnSubscribe(__ -> Log.d(TAG, "Activating project " + id))
+        .doOnSuccess(localDataStore::insertOrUpdateProject);
+  }
+
+  private Single<Project> getProjectFromLocal(String id, Throwable error) {
+    return localDataStore
+        .getProjectById(id)
+        .toSingle()
+        .doOnSubscribe(__ -> Log.d(TAG, "Failed to load project from remote", error));
```

Comment:
remote -> local?

Reply:
This was only run when the remote fetch failed. That probably wasn't too clear - pushed e54973073436e080ebb73ffba53ad23b678269e2 to make things clearer.